src/traces/scattergl/select.js

```diff
@@ -0,0 +1,70 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var subtypes = require('../scatter/subtypes');
+
+module.exports = function selectPoints(searchInfo, polygon) {
+    var cd = searchInfo.cd,
+        xa = searchInfo.xaxis,
+        ya = searchInfo.yaxis,
+        selection = [],
+        trace = cd[0].trace,
+        curveNumber = trace.index,
+        i,
+        di,
+        x,
+        y;
+
+    var scattergl = cd[0].glTrace;
+    var scene = cd[0].glTrace.scene;
+
+    // TODO: include lines? that would require per-segment line properties
+    var hasOnlyLines = (!subtypes.hasMarkers(trace) && !subtypes.hasText(trace));
+    if(trace.visible !== true || hasOnlyLines) return;
+
+    // filter out points by visible scatter ones
+    // var scatter2d = scattergl.scatter.instance
+
+    if(polygon === false) { // clear selection
+        for(i = 0; i < cd.length; i++) cd[i].dim = 0;
+    }
+    else {
+        for(i = 0; i < cd.length; i++) {
+            di = cd[i];
+            //FIXME: this affects performance for 1e6 points
+            x = xa.c2p(di.x);
+            y = ya.c2p(di.y);
```

Comment:
So... the only things you need from `scatter.calc` are `di.x` and `di.y` right? There's a whole lot that could be cut out from `scatter.calc` then - like, probably 90% of the code, though I'm not sure what fraction of time this corresponds to if `x` and `y` are the only arrays in the trace - but I bet at least removing `Axes.expand` is a decent savings.

Perhaps we could also stash these pixel positions somewhere, so `c2p` only has to be called on the first `selectPoints` after any zoom/pan. Probably `polygon.contains` is the more expensive operation much of the time, and you'd have to be careful about what conditions would invalidate these pixel values, but could be a worthwhile optimization.

It may be a bigger project than you want to take on right now, but I suspect the longer-term solution is to not create a `cd` at all (or just a dummy if you need it to pass a reference to `trace` along here) and do something more like @monfera was describing parcoords doing, and have this picking happen on the GPU. Take this part as speculation since I haven't worked with GPU code, but something like:
- stash pixel positions in x and y as arrays on the GPU during rendering, so the picker loop doesn't have to do any `c2p` calculations
- reimplement `polygon.contains` on the GPU
- pass a selection array back to the CPU for emitting events